caput/pfb.py

```diff
@@ -0,0 +1,190 @@
+
+.. autosummary::
+    :toctree: generated/
+
+    sinc_window
+    sinc_hanning
+    sinc_hamming
+
+PFB
+===
+
+.. autosummary::
+    :toctree: generated/
+
+    pfb
+    decorrelation_ratio
+"""
+# === Start Python 2/3 compatibility
+from __future__ import absolute_import, division, print_function, unicode_literals
+from future.builtins import *  # noqa  pylint: disable=W0401, W0614
+from future.builtins.disabled import *  # noqa  pylint: disable=W0401, W0614
+
+# === End Python 2/3 compatibility
+
+import numpy as np
+from scipy.interpolate import interp1d
+
+
+def sinc_window(ntap, lblock):
+    """Sinc window function.
+
+    Parameters
+    ----------
+    ntaps : integer
+        Number of taps.
+    lblock: integer
+        Length of block.
+
+    Returns
+    -------
+    window : np.ndarray[ntaps * lblock]
+    """
+    # Sampling locations of sinc function
+    X = np.linspace(-ntap / 2, ntap / 2, ntap * lblock, endpoint=False)
+
+    # np.sinc function is sin(pi*x)/pi*x, not sin(x)/x, so we can just X
+    return np.sinc(X)
+
+
+def sinc_hanning(ntap, lblock):
+    """Hanning-sinc window function.
+
+    Parameters
+    ----------
+    ntaps : integer
+        Number of taps.
+    lblock: integer
+        Length of block.
+
+    Returns
+    -------
+    window : np.ndarray[ntaps * lblock]
+    """
+
+    return sinc_window(ntap, lblock) * np.hanning(ntap * lblock)
+
+
+def sinc_hamming(ntap, lblock):
+    """Hamming-sinc window function.
+
+    Parameters
+    ----------
+    ntaps : integer
+        Number of taps.
+    lblock: integer
+        Length of block.
+
+    Returns
+    -------
+    window : np.ndarray[ntaps * lblock]
+    """
+
+    return sinc_window(ntap, lblock) * np.hamming(ntap * lblock)
+
+
+def pfb(timestream, nfreq, ntap=4, window=sinc_hamming):
+    """Perform the CHIME PFB on a timestream.
+
+    Parameters
+    ----------
+    timestream : np.ndarray
+        Timestream to process
+    nfreq : int
+        Number of frequencies we want out (probably should be odd
+        number because of Nyquist)
+    ntaps : int
+        Number of taps.
+
+    Returns
+    -------
+    pfb : np.ndarray[:, nfreq]
+        Array of PFB frequencies.
+    """
+
+    # Number of samples in a sub block
+    lblock = 2 * (nfreq - 1)
+
+    # Number of blocks
+    nblock = timestream.size // lblock - (ntap - 1)
+
+    # Initialise array for spectrum
+    spec = np.zeros((nblock, nfreq), dtype=np.complex128)
+
+    # Window function
+    w = window(ntap, lblock)
+
+    # Iterate over blocks and perform the PFB
+    for bi in range(nblock):
+        # Cut out the correct timestream section
+        ts_sec = timestream[(bi * lblock) : ((bi + ntap) * lblock)].copy()
+
+        # Perform a real FFT (with applied window function)
+        ft = np.fft.rfft(ts_sec * w)
+
+        # Choose every n-th frequency
+        spec[bi] = ft[::ntap]
+
+    return spec
+
+
+def decorrelation_ratio(ntap, lblock, window=sinc_hamming, oversample=4):
```

Comment:
Are functions mutable? I'm not sure if the best practice is to set the default to be `None` just say this corresponds to sing_hamming in the doctoring.